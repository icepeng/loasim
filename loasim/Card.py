from typing import List, Tuple
from loasim.CharacterModifier import CharacterModifier


class Card:
    def __init__(self, name: str, mdf: CharacterModifier) -> None:
        self.name = name
        self.mdf = mdf

    def get_modifier(self):
        return self.mdf


card_list = [
    Card("남겨진 바람의 절벽 (12)", CharacterModifier(crit=7)),
    Card("세상을 구하는 빛 (18)", CharacterModifier(pdamage_indep=7)),
    Card("세상을 구하는 빛 (30)", CharacterModifier(pdamage_indep=15)),
]

card_dict = {card.name: card for card in card_list}


class CardManager:
    def __init__(self, *equipped_list: List[str]) -> None:
        self.equipped_list = equipped_list

    def get_static_modifier(self):
        mdf = CharacterModifier()
        for name in self.equipped_list:
            card = card_dict.get(name)
            if card is None:
                raise Exception(f"{name} is not an card")
            mdf = mdf + card.get_modifier()
        return mdf
