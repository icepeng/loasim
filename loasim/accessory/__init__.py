from loasim.accessory.builtin import lostark_accessory_repository
from loasim.accessory.base import AccessoryState
