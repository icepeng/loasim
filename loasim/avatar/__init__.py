from loasim.avatar.base import AvatarState
from loasim.avatar.builtin import lostark_avatar_repository
