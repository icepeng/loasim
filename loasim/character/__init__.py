from loasim.character.base import AccessoryState, AvatarState, Character, GearState
