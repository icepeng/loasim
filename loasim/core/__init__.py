from loasim.core.buff import (
    BuffManager,
    OnoffBuff,
    SkillBuff,
    StackBuff,
    StatBuff,
    StaticBuff,
)
from loasim.core.enemy import Enemy
from loasim.core.skill import Skill, SkillSpecification, Tripod
from loasim.core.stat import InternalStat, Stat
