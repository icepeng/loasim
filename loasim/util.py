from typing import Any, Dict

import yaml


def save_config(filename: str, obj: Dict[str, Any]):
    with open(
        filename,
        mode="w",
        encoding="utf8",
    ) as f:
        yaml.safe_dump(obj, f, allow_unicode=True, sort_keys=False)


def load_config(filename: str):
    with open(
        filename,
        encoding="utf8",
    ) as f:
        return yaml.safe_load(f)
