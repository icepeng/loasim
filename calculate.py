from collections import defaultdict
from typing import List
from loasim.core import Skill


def calculate(dealcycle: List[Skill], cycle_time, enemy):
    total_damage = 0
    damage_dict = defaultdict(float)
    for sk in dealcycle:
        damage = sk.get_damage(enemy)
        total_damage += damage
        damage_dict[sk.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(f"dps: {total_damage / cycle_time:,.0f}")
