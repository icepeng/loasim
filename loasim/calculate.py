from loguru import logger

from collections import defaultdict
from typing import Any, Callable, List, Dict, Literal, Tuple

from pydantic import BaseModel

from loasim.core import Skill, Enemy, BuffManager, InternalStat
from loasim.core.buff import Buff
from loasim.core.stat import Stat
from loasim.setitem import lostark_setitem_repository
from loasim.card import lostark_default_card_repository
from loasim.engraving import lostark_engraving_repository


class DealCycle(BaseModel):
    buff_list: List[Buff]
    skill_list: List[Tuple[Skill, Dict[str, Any], Literal["back", "head"] | None]]
    cycle_time: float


def calculate(
    internal_stat: InternalStat,
    weapon_pdamage: float,
    setitem_state: List[Tuple[str, int, int]],
    engraving_state: List[Tuple[str, int]],
    card_state: List[str],
    generator: Callable[[InternalStat], DealCycle],
    enemy: Enemy,
):
    deal_cycle = generator(internal_stat)
    base_stat = internal_stat.get_stat()
    weapon_stat = Stat(pdamage=weapon_pdamage)  # 무기 품질
    basis_stat = (
        base_stat
        + weapon_stat
        + lostark_setitem_repository.get_stat(setitem_state)
        + lostark_default_card_repository.get_stat(card_state)
    )

    logger.info(base_stat)
    logger.info(basis_stat)

    engraving_buffs = lostark_engraving_repository.get_buffs(engraving_state)
    buff_manager = BuffManager(engraving_buffs + deal_cycle.buff_list)

    total_damage = 0.0
    damage_dict: defaultdict[str, float] = defaultdict(float)
    for skill, buff_state, backhead in deal_cycle.skill_list:
        buff_stat = buff_manager.get_stat(buff_state, skill)
        damage = skill.get_damage(enemy, backhead, basis_stat + buff_stat)
        total_damage += damage
        damage_dict[skill.name] += damage

    for sk, dmg in damage_dict.items():
        print(sk, f"{dmg / total_damage *100:,.3f}%")
    print(f"total damage: {total_damage:,.0f}")
    print(deal_cycle.cycle_time)
    print(f"dps: {total_damage / deal_cycle.cycle_time:,.0f}")
