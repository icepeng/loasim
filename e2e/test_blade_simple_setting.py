from os import path

from loasim.calculate import DealCycle, calculate
from loasim.character import Character
from loasim.core import Enemy
from loasim.util import load_config


def calculate_blade():
    character = Character.parse_obj(
        load_config(path.join(path.dirname(__file__), "configs", "blade_character.yml"))
    )
    deal_cycle = DealCycle.parse_obj(
        load_config(path.join(path.dirname(__file__), "configs", "blade_dealcycle.yml"))
    )

    return calculate(
        character=character,
        enemy=Enemy(armor=6000, reduction=23),
        deal_cycle=deal_cycle,
    )


def test_blade_simple_setting(snapshot: float):
    assert calculate_blade() == snapshot


if __name__ == "__main__":
    calculate_blade()
