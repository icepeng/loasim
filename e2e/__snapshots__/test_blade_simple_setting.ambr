# name: test_blade_simple_setting
  5517682.218923723
---
